### PR TITLE
Introduce `Vector` type in SQL

### PR DESCRIPTION
**Pull request by hatyo**
This PR introduces native support for vector data types in the RelationalFDB, enabling storage and querying of fixed-size numerical vectors commonly used in machine learning and similarity search applications through SQL.

The vector type represents a fixed-dimensional array of floating-point numbers with configurable precision. Vectors are declared using the VECTOR(dimension, precision) syntax:

```sql
  CREATE TABLE embeddings (
      id BIGINT,
      embedding VECTOR(128, FLOAT),
      PRIMARY KEY(id)
  );
```
supported precision types are:
  - `HALF` - 16-bit half-precision (2 bytes per element)
  - `FLOAT` - 32-bit single-precision (4 bytes per element)
  - `DOUBLE` - 64-bit double-precision (8 bytes per element)

**Key Features**

####  1. Type Conversions:
Cast numeric arrays to vectors with automatic type promotion:

```sql
  SELECT CAST([1, 2, 3] AS VECTOR(3, FLOAT)) AS vec;
  SELECT CAST([1.5, 2.5, 3.5] AS VECTOR(3, HALF)) AS vec;
```

Supports casting from:
- `INTEGER`, `BIGINT`, `FLOAT`, `DOUBLE` arrays.

#### 2. Query Operations:

  - Equality/Inequality: =, !=
  - NULL handling: `IS NULL`, `IS NOT NULL`
  - NULL-safe comparisons: `IS DISTINCT FROM`, `IS NOT DISTINCT FROM`

Similar to other primitive types, vectors can be used in `struct`, and (nullable) `array`s.

#### 3. JDBC Integration

Vectors are inserted using prepared statements with vector objects:

```java
  PreparedStatement stmt = connection.prepareStatement(
      "INSERT INTO embeddings VALUES (?, ?)");
  stmt.setObject(2, new FloatRealVector(new float[]{0.5f, 1.2f, -0.8f}));
```

#### 4. YAML Testing support:

Introduced new YAML tags to enable defining and validating vectors via `!vXX [values]` syntax where:

* `!v16` represents a `HALF` precision vector parameter
* `!v32` represents a `FLOAT` precision vector parameter
* `!v64` represents a `DOUBLE` precision vector parameter

this fixes https://github.com/FoundationDB/fdb-record-layer/issues/3693, it also rectifies incorrect tests as described in #3695, and (partial) fixes for #3665 as well.

### CODE REVIEW

File: yaml-tests/src/test/resources/vector.yamsql [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+---
+options:
+  supported_version: !current_version
+---
+schema_template:
+    create table tHalfVector(a bigint, b vector(3, half), primary key(a))
+    create table tFloatVector(a bigint, b vector(3, float), primary key(a))
+    create table tDoubleVector(a bigint, b vector(3, double), primary key(a))
```
Comment (arnaud-lacurie) [inline yaml-tests/src/test/resources/vector.yamsql:28]: I don't see why it wouldn't work, but can you add a test that has multiple vectors in the same table? 
Of the same type, of different types? 
Can you add an array of vectors, and a struct with vectors in it?

File: fdb-relational-grpc/src/main/proto/grpc/relational/jdbc/v1/jdbc.proto [hunk 2]
```diff
@@ -108,11 +108,23 @@ message Parameters {
...
   // Reuse the column type here. It has much of what we need carrying
   // across parameters as well as null support, etc.
-  Column parameter = 2;
-  Type type = 3;
+  Column parameter = 2; // deprecated
+  Type type = 3; // deprecated
+  ParameterMetadata metadata = 4;
+}
+
+message ParameterMetadata {
```
Comment (arnaud-lacurie) [inline fdb-relational-grpc/src/main/proto/grpc/relational/jdbc/v1/jdbc.proto:119]: I'm not sure I understand the reason behind putting everything in metadata.

Why not keep `java_sql_types_code` and `type` since they are in metadata anyway, and only deprecate `parameter`?

File: fdb-relational-grpc/src/main/proto/grpc/relational/jdbc/v1/column.proto [hunk 3]
```diff
@@ -130,3 +132,8 @@ message EnumMetadata {
   string name = 1;
   repeated string values = 2;
 }
+
+message VectorMetadata {
+  int32 precision = 1;
```
Comment (arnaud-lacurie) [inline fdb-relational-grpc/src/main/proto/grpc/relational/jdbc/v1/column.proto:137]: Do we need to have a base-type here as well? Or do we anticipate that we will never return vectors of int, but merely use them internally.

File: yaml-tests/src/test/resources/vector.yamsql (hunk 1 above)
Comment (hatyo) [reply yaml-tests/src/test/resources/vector.yamsql:28]: This is complex to do with prepared API bridge to YAML and may require special YAML tags I think to support what you're mentioning. I will check what we can do here.

Comment (hatyo) [reply yaml-tests/src/test/resources/vector.yamsql:28]: .... I ended up adding tests for defining vectors inside arrays, and structs, and I also added support for defining nullable vector arrays through:
- JDBC/GRPC
- DynamicMessage --> Row[Struct|Array]
I am not 100% sure this addresses all potential bugs in complex scenarios, such as adding a nullable vector array inside a struct inside another array and so on, but there are now few more tests with vectors are inside complex structures inside `vector.yamsql` and they run successfully.

File: fdb-relational-grpc/src/main/proto/grpc/relational/jdbc/v1/jdbc.proto (hunk 2 above)
Comment (hatyo) [reply fdb-relational-grpc/src/main/proto/grpc/relational/jdbc/v1/jdbc.proto:119]: I incorrectly documented `parameter` as `deprecated`, which caused the confusion. The `Parameter` message now has this structure:

```proto
// A parameter has java sql type as well as actual parameter.
message Parameter {
  int32 java_sql_types_code = 1; // deprecated
  Column parameter = 2;
  Type type = 3; // deprecated
  ParameterMetadata metadata = 4;
}
```

in two sprints from now, we'll clean it up to be:

```proto
message Parameter {
  Column parameter = 2;
  ParameterMetadata metadata = 4;
}
```

For now, I just removed the `deprecated` comment added to the `parameter` field.

File: fdb-relational-grpc/src/main/proto/grpc/relational/jdbc/v1/column.proto (hunk 3 above)
Comment (hatyo) [reply fdb-relational-grpc/src/main/proto/grpc/relational/jdbc/v1/column.proto:137]: That's a good point, I think this is ok for now, in the future, we could amend this message and add a new field for example, indicating whether the element is a decimal or floating-point:

```proto
message VectorMetadata {
  int32 precision = 1;
  int32 dimensions = 2;
  bool isDecimal = 3;
}
```

So, representing for example a 128-dimensions vector of `INT8` would correspond to this message `{precision=8, dimensions=128, isDecimal =false}`, whereas an existing message like `{precision=16, dimensions=768}` continues to be parsed as `VECTOR(768, HALF)`.

File: docs/sphinx/source/reference/sql_types.rst [hunk 4]
```diff
@@ -72,14 +72,172 @@ In this example, `c` is an array, and each record within the array is a struct o
...
+.. code-block:: java
+
+    // Java JDBC example
+    PreparedStatement stmt = connection.prepareStatement(
+        "INSERT INTO embeddings VALUES (?, ?)");
+    stmt.setLong(1, 1);
+    stmt.setObject(2, new FloatRealVector(new float[]{0.5f, 1.2f, -0.8f}));
+    stmt.executeUpdate();
+
+For documentation and testing purposes, the examples below use a special test DSL syntax :sql:`!! !vXX [values] !!` to represent prepared statement parameters:
```
Comment (arnaud-lacurie) [inline docs/sphinx/source/reference/sql_types.rst]: This probably belongs in a separate "developer-guide" section of the doc, focused on how to write tests in the yamsql framework. I'm fine if that stays here until we create that though.

File: docs/sphinx/source/reference/sql_types.rst [hunk 5]
```diff
@@ -72,14 +72,172 @@ In this example, `c` is an array, and each record within the array is a struct o
...
+
+Vectors can be selected and compared like other column types. When comparing vectors in WHERE clauses, you would typically use prepared statement parameters in your Java/JDBC code:
+
+.. code-block:: sql
+
+    -- Select vectors
+    SELECT embedding FROM embeddings WHERE id = 1;
+
+    -- Compare vectors for equality (using test DSL syntax; in actual code use PreparedStatement)
+    SELECT id FROM embeddings WHERE embedding = !! !v32 [0.5, 1.2, -0.8] !!;
```
Comment (arnaud-lacurie) [inline docs/sphinx/source/reference/sql_types.rst]: If we want to restrict DSL syntax to the developer guide, you could write a cast here instead.

Comment (arnaud-lacurie) [reply docs/sphinx/source/reference/sql_types.rst]: Or simply write `embedding = ?`

File: docs/sphinx/source/reference/sql_types.rst [hunk 6]
```diff
@@ -72,14 +72,172 @@ In this example, `c` is an array, and each record within the array is a struct o
...
+
+Supported Operations
+====================
+
+The following operations are supported on vector types:
+
+* **Equality comparison** (:sql:`=`, :sql:`!=`)
+* **NULL checks** (:sql:`IS NULL`, :sql:`IS NOT NULL`)
+* **NULL-safe comparison** (:sql:`IS DISTINCT FROM`, :sql:`IS NOT DISTINCT FROM`)
+* **CAST from numeric arrays** to vectors
```
Comment (arnaud-lacurie) [inline docs/sphinx/source/reference/sql_types.rst:234]: Should there be a mention here of the fact that you cannot create value indexes on vectors? 
We also probably want to mention that you cannot ORDER BY a vector field at this point. 
Can we group by a vector field?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/expressions/Comparisons.java [hunk 7]
```diff
@@ -224,6 +225,8 @@ public static Object toClassWithRealEquals(@Nonnull Object obj) {
             return obj;
         } else if (obj instanceof List) {
             return obj;
+        } else if (obj instanceof RealVector) {
+            return obj;
```
Comment (arnaud-lacurie) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/expressions/Comparisons.java:229]: It's probably a good idea to address this Teamscale warning.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/typing/Type.java [hunk 8]
```diff
@@ -1176,6 +1215,157 @@ public boolean equals(final Object other) {
         }
     }
 
+    class Vector implements Type {
```
Comment (arnaud-lacurie) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/typing/Type.java]: Can you make that class final?

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/typing/Type.java [hunk 9]
```diff
@@ -1176,6 +1215,157 @@ public boolean equals(final Object other) {
...
+            final PVectorType.Builder vectorTypeBuilder = PVectorType.newBuilder()
+                    .setIsNullable(isNullable)
+                    .setDimensions(dimensions)
+                    .setPrecision(precision);
+            return vectorTypeBuilder.build();
+        }
+
+        @Nonnull
+        @SuppressWarnings("PMD.ReplaceVectorWithList")
+        public static Vector fromProto(@Nonnull final PlanSerializationContext serializationContext,
```
Comment (arnaud-lacurie) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/typing/Type.java]: Should `serializationContext` be removed?

File: yaml-tests/src/test/resources/vector.yamsql [hunk 10]
```diff
@@ -0,0 +1,449 @@
...
+      - query: select b is distinct from !! !v64 [1.3, 1.4, 1.6] !! from tDoubleVector where a = 100
+      - result: [{false}]
+    -
+      - query: select b is distinct from !! !v64 [9.9, 9.9, 9.9] !! from tDoubleVector where a = 100
+      - result: [{true}]
+    -
+      - query: select b is distinct from null from tDoubleVector where a = 200
+      - result: [{false}]
+    -
+      - query: select b is distinct from !! !v64 [1.3, 1.4, 1.6] !! from tDoubleVector where a = 200
```
Comment (arnaud-lacurie) [inline yaml-tests/src/test/resources/vector.yamsql:311]: You may want to add distinct / not distinct tests that check nulls that are in the table, and not just null literals.

File: yaml-tests/src/main/java/com/apple/foundationdb/relational/yamltests/server/ExternalServer.java [hunk 11]
```diff
@@ -44,7 +44,7 @@ public class ExternalServer {
 
     private static final Logger logger = LogManager.getLogger(ExternalServer.class);
     public static final String EXTERNAL_SERVER_PROPERTY_NAME = "yaml_testing_external_server";
-    private static final boolean SAVE_SERVER_OUTPUT = false;
+    private static final boolean SAVE_SERVER_OUTPUT = true;
```
Comment (arnaud-lacurie) [inline yaml-tests/src/main/java/com/apple/foundationdb/relational/yamltests/server/ExternalServer.java]: Does this need to be set back to false?

File: fdb-relational-server/src/main/resources/log4j2.xml [hunk 12]
```diff
@@ -19,17 +19,17 @@
  limitations under the License.
 -->
 
-<Configuration status="INFO">
+<Configuration status="DEBUG">
```
Comment (arnaud-lacurie) [inline fdb-relational-server/src/main/resources/log4j2.xml]: This needs to be set back to INFO

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/RelOpValue.java [hunk 13]
```diff
@@ -1042,6 +1042,24 @@ private enum BinaryPhysicalOperator {
...
+        LTE_VEC_NULL(Comparisons.Type.LESS_THAN_OR_EQUALS, Type.TypeCode.VECTOR, Type.TypeCode.NULL, (l, r) -> null),
+        LTE_NULL_VEC(Comparisons.Type.LESS_THAN_OR_EQUALS, Type.TypeCode.NULL, Type.TypeCode.VECTOR, (l, r) -> null),
+        GT_VEC_NULL(Comparisons.Type.GREATER_THAN, Type.TypeCode.VECTOR, Type.TypeCode.NULL, (l, r) -> null),
+        GT_NULL_VEC(Comparisons.Type.GREATER_THAN, Type.TypeCode.NULL, Type.TypeCode.VECTOR, (l, r) -> null),
+        GTE_VEC_NULL(Comparisons.Type.GREATER_THAN_OR_EQUALS, Type.TypeCode.VECTOR, Type.TypeCode.NULL, (l, r) -> null),
+        GTE_NULL_VEC(Comparisons.Type.GREATER_THAN_OR_EQUALS, Type.TypeCode.NULL, Type.TypeCode.VECTOR, (l, r) -> null),
+        IS_DISTINCT_FROM_NULL_VEC(Comparisons.Type.IS_DISTINCT_FROM, Type.TypeCode.NULL, Type.TypeCode.VECTOR, (l, r) -> Objects.nonNull(r)),
+        IS_DISTINCT_FROM_VEC_NULL(Comparisons.Type.IS_DISTINCT_FROM, Type.TypeCode.VECTOR, Type.TypeCode.NULL, (l, r) -> Objects.nonNull(l)),
+        IS_DISTINCT_FROM_VEC_VEC(Comparisons.Type.IS_DISTINCT_FROM, Type.TypeCode.VECTOR, Type.TypeCode.VECTOR, (l, r) -> !l.equals(r)),
```
Comment (arnaud-lacurie) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/RelOpValue.java:1061]: This would similarly NPE if one of them is null at runtime.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/RelOpValue.java [hunk 14]
```diff
@@ -1042,6 +1042,24 @@ private enum BinaryPhysicalOperator {
 
         NOT_DISTINCT_FROM_NN(Comparisons.Type.NOT_DISTINCT_FROM, Type.TypeCode.NULL, Type.TypeCode.NULL, (l, r) -> true),
 
+        EQ_VEC_VEC(Comparisons.Type.EQUALS, Type.TypeCode.VECTOR, Type.TypeCode.VECTOR, Objects::equals),
+        NEQ_VEC_VEC(Comparisons.Type.NOT_EQUALS, Type.TypeCode.VECTOR, Type.TypeCode.VECTOR, (l, r) -> !l.equals(r)),
```
Comment (arnaud-lacurie) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/RelOpValue.java:1048]: This would NPE if `l` is null at runtime.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/RelOpValue.java [hunk 15]
```diff
@@ -1042,6 +1042,24 @@ private enum BinaryPhysicalOperator {
 
         NOT_DISTINCT_FROM_NN(Comparisons.Type.NOT_DISTINCT_FROM, Type.TypeCode.NULL, Type.TypeCode.NULL, (l, r) -> true),
 
+        EQ_VEC_VEC(Comparisons.Type.EQUALS, Type.TypeCode.VECTOR, Type.TypeCode.VECTOR, Objects::equals),
```
Comment (arnaud-lacurie) [inline fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/RelOpValue.java:1045]: Do we need EQ_VEC_NULL and EQ_NULL_VEC as well?

**Review by arnaud-lacurie — CHANGES_REQUESTED** [review]

File: docs/sphinx/source/reference/sql_types.rst (hunk 4 above)
Comment (hatyo) [reply docs/sphinx/source/reference/sql_types.rst]: This is already resolved in the most recent commit.

File: docs/sphinx/source/reference/sql_types.rst (hunk 5 above)
Comment (hatyo) [reply docs/sphinx/source/reference/sql_types.rst]: This is already resolved in the most recent commit. `CAST` is now used instead, with explanation in the documentation why this is the case.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/RelOpValue.java (hunk 15 above)
Comment (hatyo) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/RelOpValue.java:1045]: Nice catch, I thought I already added that, added now with tests.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/RelOpValue.java (hunk 14 above)
Comment (hatyo) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/RelOpValue.java:1048]: That can not happen. If `l` is `null` a different implementation will be called (`NEQ_NULL_VEC`)

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/RelOpValue.java (hunk 13 above)
Comment (hatyo) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/values/RelOpValue.java:1061]: That can not happen. If `l` is `null` a different implementation will be called.

File: yaml-tests/src/test/resources/vector.yamsql (hunk 10 above)
Comment (hatyo) [reply yaml-tests/src/test/resources/vector.yamsql:311]: The same logic will kick in regardless to where the `null` originate from, I may add these tests later on not in this PR.

File: fdb-relational-server/src/main/resources/log4j2.xml (hunk 12 above)
Comment (hatyo) [reply fdb-relational-server/src/main/resources/log4j2.xml]: Done.

File: yaml-tests/src/main/java/com/apple/foundationdb/relational/yamltests/server/ExternalServer.java (hunk 11 above)
Comment (hatyo) [reply yaml-tests/src/main/java/com/apple/foundationdb/relational/yamltests/server/ExternalServer.java]: Yep, done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/typing/Type.java (hunk 8 above)
Comment (hatyo) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/typing/Type.java]: Done. We should probably do that for other types as well.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/typing/Type.java (hunk 9 above)
Comment (hatyo) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/typing/Type.java]: Done.

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/expressions/Comparisons.java (hunk 7 above)
Comment (hatyo) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/expressions/Comparisons.java:229]: ok. done.

File: docs/sphinx/source/reference/sql_types.rst (hunk 6 above)
Comment (hatyo) [reply docs/sphinx/source/reference/sql_types.rst:234]: I haven't tested any of that (yet), but I can get to it later (not important for now, but nice to cover for completeness).

**Review by arnaud-lacurie — APPROVED** [review]

File: fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/expressions/Comparisons.java (hunk 7 above)
Comment (hatyo) [reply fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/expressions/Comparisons.java:229]: the refactoring changed the business logic and broke certain plan hash code calculations in a subtle way; this is due to the way these comparisons are laid out, giving special handling of specific types that implement interfaces existing in branches further down below.

One test the broke had a comparison against a deserialized protobuf `enum` value, the deserialized protobuf enum is of type `com.google.protobuf.ProtocolMessageEnum` which extends `Internal.EnumLite`, so passing an object of that type to the `if` below will take `branch 1` and return the enum's number to comparison and of calculating hash code.

In Java spec, `enum`s intrinsically map to `Enum` class which [implements `Comparable`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Enum.html#:~:text=Serializable%2C-,Comparable%3CE%3E,-%2C%20Constable), and so passing the same object to the `if` below will end up taking`branch 2` and returning the enum object itself for calculating the hash code, if we were to reorganize the code such that all the `if` branches sharing the return branch are put together as previously done [0].

Therefore, I reverted the code to how it was before.

```java
if (obj instanceof ByteString) {
      return obj;
} else if (obj instanceof byte[]) {
      return ByteString.copyFrom((byte[])obj);
} else if (obj instanceof Internal.EnumLite) { 
      return ((Internal.EnumLite)obj).getNumber(); <---------- branch 1
} else if (obj instanceof Comparable) { 
      return obj; <--------------- branch 2
} else if (obj instanceof List) {
      return obj;
} else if (obj instanceof RealVector) {
      return obj;
} else {
      throw new RecordCoreException("Tried to compare non-comparable object " + obj.getClass());
}
```

[0]
```java
if (obj instanceof ByteString || obj instanceof Comparable 
    || obj instanceof List || obj instanceof RealVector) { 
      return obj; <---- branch 2
} else if (obj instanceof byte[]) {
     return ByteString.copyFrom((byte[])obj);
} else if (obj instanceof Internal.EnumLite) { 
     return ((Internal.EnumLite)obj).getNumber(); <---------- branch 1
} else {
     throw new RecordCoreException("Tried to compare non-comparable object " + obj.getClass());
}
```
